### PR TITLE
AM-55-store-password-history-for-user

### PR DESCRIPTION
**Pull request by farmborough**
## :id: Reference related issue. 
Fixes [AM-55](https://graviteecommunity.atlassian.net/browse/AM-55).

## :pencil2: 

This is just the persistence functionality of password history. I've left the encryption responsibility to the PasswordHistoryService.

## :memo: 

All mongo and JDBC repo tests have been successfully run.

### CODE REVIEW

File: gravitee-am-repository/gravitee-am-repository-tests/src/test/java/io/gravitee/am/repository/management/api/PasswordHistoryRepositoryTest.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        testObserver.awaitTerminalEvent();
+        testObserver.assertComplete();
+        testObserver.assertNoErrors();
+        assertEqualsTo(history, testObserver);
+    }
+
+    @Test
+    public void shouldDeletePasswordHistory() {
+        var created = repository.create(buildPasswordHistory()).blockingGet();
+        var testObserver = repository.delete(created.getId()).test();
```
[the remark this answers is not in the document]
Comment (farmborough) [reply gravitee-am-repository/gravitee-am-repository-tests/src/test/java/io/gravitee/am/repository/management/api/PasswordHistoryRepositoryTest.java:48]: It won't hurt to add, although `create` is also tested so it's redundant here.

File: gravitee-am-repository/gravitee-am-repository-jdbc/src/main/resources/liquibase/changelogs/v3_20_0/3.20.0-add-password-histories-table.yml [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+            tableName: password_histories
+            unique: false
+
+        - createIndex:
+            columns:
+              - column:
+                  name: user_id
+            indexName: idx_password_histories_user
+            tableName: password_histories
+            unique: false
```
[the remark this answers is not in the document]
Comment (farmborough) [reply gravitee-am-repository/gravitee-am-repository-jdbc/src/main/resources/liquibase/changelogs/v3_20_0/3.20.0-add-password-histories-table.yml:54]: I wasn't sure about indices, TBH. Good to know this.

File: gravitee-am-repository/gravitee-am-repository-jdbc/src/main/java/io/gravitee/am/repository/jdbc/management/api/JdbcPasswordHistoryRepository.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+        return monoToSingle(updateAction.as(trx::transactional)).flatMap(i -> this.findById(item.getId()).toSingle())
+                .doOnError(error -> LOGGER.error("unable to update password history with id {}", item.getId(), error));
+    }
+
+    @Override
+    public Completable delete(String id) {
+        LOGGER.debug("delete password history with id {}", id);
+        TransactionalOperator trx = TransactionalOperator.create(tm);
+        return monoToCompletable(template.delete(JdbcI18nDictionary.class)
```
[the remark this answers is not in the document]
Comment (farmborough) [reply gravitee-am-repository/gravitee-am-repository-jdbc/src/main/java/io/gravitee/am/repository/jdbc/management/api/JdbcPasswordHistoryRepository.java:122]: I'm surprised the test still worked.

File: gravitee-am-repository/gravitee-am-repository-tests/src/test/java/io/gravitee/am/repository/management/api/PasswordHistoryRepositoryTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (farmborough) [reply gravitee-am-repository/gravitee-am-repository-tests/src/test/java/io/gravitee/am/repository/management/api/PasswordHistoryRepositoryTest.java:48]: What about `assertNotNull(created.getId()`?

**Comment by farmborough** [comment]
Just realised I should have added ordering by date to the `findUserHistory` method.

**Review by leleueri — APPROVED** [review]

**Pull request merged**